### ISSUE TITLE
String literal reported as "is not defined" generating `reportUndefinedVariable`

### ISSUE THREAD

**Issue by aslpavel** [microsoft/pyright#8649]
String literal is reported as undefined variable producing `reportUndefinedVariable`

**Code or Screenshots**
Minimal example

```python
mem = memoryview(b"0123456789").cast("B", (2, 5))
```

Generated error:
```
$ pyright issue.py
/home/user/issue.py
  /home/user/issue.py:1:39 - error: "B" is not defined (reportUndefinedVariable)
1 error, 0 warnings, 0 informations
```

**Version**
`pyright 1.1.373`

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so this isn't a bug.

The first argument to a `cast` call is interpreted as a type expression. For details, refer to [this section](https://typing.readthedocs.io/en/latest/spec/directives.html#cast) of the typing spec. Quoted type expressions are treated as forward references, but `B` is not defined in your example.

**Comment by aslpavel** [comment]
Could you please clarify what you mean here, it is not `typing.cast` it is a method on `memoryview` type which is define in typeshed as:
```python
_IntegerFormats: TypeAlias = Literal[
    "b", "B", "@b", "@B", "h", "H", "@h", "@H", "i", "I", "@i", "@I", "l", "L", "@l", "@L", "q", "Q", "@q", "@Q", "P", "@P"
]

@final
class memoryview(Sequence[_I]):
    @overload
    def cast(self, format: Literal["c", "@c"], shape: list[int] | tuple[int, ...] = ...) -> memoryview[bytes]: ...
    @overload
    def cast(self, format: Literal["f", "@f", "d", "@d"], shape: list[int] | tuple[int, ...] = ...) -> memoryview[float]: ...
    @overload
    def cast(self, format: Literal["?"], shape: list[int] | tuple[int, ...] = ...) -> memoryview[bool]: ...
    @overload
    def cast(self, format: _IntegerFormats, shape: list[int] | tuple[int, ...] = ...) -> memoryview: ...
```

which can clearly accepts `"B"` literal and should infer type as `memoryview[int]`

**Comment by erictraut** [comment]
Ah, I misread your sample. I'll reopen and investigate further.

**Comment by erictraut** [comment]
Pyright's logic was making the same mistake that I was when I (too hastily) read your code sample and assumed that `cast` was referring to `typing.cast`. This is indeed a bug, and it will be fixed in the next release. Thanks for reporting the issue.

**Comment by erictraut** [comment]
This is included in pyright 1.1.375.

**Comment by stefanos82** [comment]
I have a related behavior with latest version `1.1.376`; I decided to play with various examples from https://github.com/fluentpython/example-code-2e/ 

The issue I face took place with https://github.com/fluentpython/example-code-2e/blob/master/11-pythonic-obj/vector2d_v3.py

The error is this:

```
$ pyright vector2d_v3.py                                                                                                                                     
/home/stefanos/development/FluentPython2nd/vector2d/vector2d_v3.py                                                                                                                             
  /home/stefanos/development/FluentPython2nd/vector2d/vector2d_v3.py:150:16 - error: No overloads for "cast" match the provided arguments (reportCallIssue)                                    
  /home/stefanos/development/FluentPython2nd/vector2d/vector2d_v3.py:150:44 - error: Argument of type "str" cannot be assigned to parameter "format" of type "_IntegerFormats" in function "cast"                                                                                                                                                                                             
    Type "str" is incompatible with type "_IntegerFormats"                                                                                                                                     
      "str" is incompatible with type "Literal['b']"                                                                                                                                           
      "str" is incompatible with type "Literal['B']"                                                                                                                                           
      "str" is incompatible with type "Literal['@b']"                                                                                                                                          
      "str" is incompatible with type "Literal['@B']"                                                                                                                                          
      "str" is incompatible with type "Literal['h']"                                                                                                                                           
      "str" is incompatible with type "Literal['H']"                                                                                                                                           
      "str" is incompatible with type "Literal['@h']"                                                                                                                                          
    ... (reportArgumentType)                                                                                                                                                                   
2 errors, 0 warnings, 0 informations
```

For some reason, `typecode = chr(octets[0])` that contains `d` of type `string`, the cast from `memv = memoryview(octets[1:]).cast(typecode)` interprets it as plain `d`...if I replace `typecode` with `'d'`, it works as expected.

**Comment by erictraut** [comment]
@stefanos82, this issue has been resolved and closed. If you think you have found a bug, please open a new issue and supply a self-contained minimal code sample that demonstrates the issue.

**Comment by stefanos82** [comment]
> @stefanos82, this issue has been resolved and closed. If you think you have found a bug, please open a new issue and supply a self-contained minimal code sample that demonstrates the issue.

Done! See #8777
